examples/ros/src/smarts_ros/scripts/ros_driver.py

```diff
@@ -0,0 +1,523 @@
+            agent.is_wrong_way = agent_obs.events.wrong_way
+            agent.is_off_route = agent_obs.events.off_route
+            agent.is_off_road = agent_obs.events.off_road
+            agent.is_on_shoulder = agent_obs.events.on_shoulder
+            agent.is_not_moving = agent_obs.events.not_moving
+            agents.agents.append(agent)
+        self._agents_publisher.publish(agents)
+
+    def _do_agents(self, observations: Dict[str, Observation]) -> Dict[str, Any]:
+        with self._control_lock:
+            actions = {
+                agent_id: self._agents[agent_id].act(agent_obs)
+                for agent_id, agent_obs in observations.items()
+            }
+            for agent_id, agent in self._agents_to_add.items():
+                spec, mission = agent[0], agent[1]
+                assert agent_id not in self._agents
+                self._agents[agent_id] = spec.build_agent()
+                self._smarts.add_agent_with_mission(agent_id, spec.interface, mission)
+            self._agents_to_add = {}
+            return actions
+
+    def _check_reset(self) -> Dict[str, Observation]:
+        with self._control_lock:
+            if self._reset_smarts:
+                rospy.loginfo(f"resetting SMARTS w/ scenario={self._scenario_path}")
+                self._reset_smarts = False
+                if self._scenario_path:
+                    observations = self._smarts.reset(Scenario(self._scenario_path))
+                    self._last_step_time = None
+                    return observations
+                return {}
+        return None
+
+    def run_forever(self):
+        if not self._state_publisher:
+            raise Exception("must call setup_ros() first.")
+        if not self._smarts:
+            raise Exception("must call setup_smarts() first.")
+        warned_scenario = False
+        observations = {}
+        step_delta = None
+        if self._target_freq:
+            rate = rospy.Rate(self._target_freq)
+        rospy.loginfo(f"starting to spin")
+        try:
+            while not rospy.is_shutdown():
+
+                obs = self._check_reset()
+                if not self._scenario_path:
+                    if not warned_scenario:
+                        rospy.loginfo("waiting for scenario on control channel...")
+                        warned_scenario = True
+                    elif self._last_step_time:
+                        rospy.loginfo("no more scenarios.  exiting...")
+                        break
+                    continue
+                elif obs is not None:
+                    observations = obs
+
+                actions = self._do_agents(observations)
+
+                if self._last_step_time:
+                    step_delta = rospy.get_time() - self._last_step_time
+                self._last_step_time = rospy.get_time()
+
+                self._update_smarts_state(step_delta)
+
+                observations, _, dones, _ = self._smarts.step(actions, step_delta)
```

Comment:
Yeah, I see your point and was just about to change it, but the alternative is a one-line function that has 2 parameters and returns a tuple.  Because of the black formatting and the type hints, that one line function becomes 4 and ends up looking very "ugly" to me.  

But I'm still willing to change it if you prefer that.